frame/nis/src/lib.rs

```diff
@@ -689,13 +708,50 @@ pub mod pallet {
 			receipt.proportion.saturating_reduce(proportion);
 			summary.proportion_owed.saturating_reduce(proportion);
 
-			T::Currency::transfer(&our_account, &who, amount, AllowDeath)
-				.map_err(|_| Error::<T>::Unfunded)?;
-
 			let dropped = receipt.proportion.is_zero();
+
+			if amount > on_hold {
+				T::Currency::unreserve_named(&T::ReserveId::get(), &who, on_hold);
+				let deficit = amount - on_hold;
+				// Try to transfer deficit from pot to receipt owner.
+				summary.receipts_on_hold.saturating_reduce(on_hold);
+				on_hold = Zero::zero();
+				T::Currency::transfer(&our_account, &who, deficit, AllowDeath)
+					.map_err(|_| Error::<T>::Unfunded)?;
+			} else {
+				T::Currency::unreserve_named(&T::ReserveId::get(), &who, amount);
+				on_hold.saturating_reduce(amount);
+				summary.receipts_on_hold.saturating_reduce(amount);
+				if dropped && !on_hold.is_zero() {
+					// Reclaim any remainder:
+					// Transfer `excess` to the pot if we have now fully compensated for the
+					// receipt.
+					//
+					// This will legitimately fail if there is no pot account in existance.
+					// There's nothing we can do about this so we just swallow the error.
+					/*let _ = T::Currency::repatriate_reserved_named(
+						&T::ReserveId::get(),
+						&who,
+						&our_account,
+						excess,
+						BalanceStatus::Free,
+					).defensive();*/
```

Comment:
```suggestion
					// This code is not ideal and could fail in the second phase leaving
					// the system in an invalid state. It can be fixed properly with the
					// new API in https://github.com/paritytech/substrate/pull/12951
					// 
					// Below is what it should look like then:
					//	let _ = T::Currency::repatriate_reserved_named(
					//		&T::ReserveId::get(),
					//		&who,
					//		&our_account,
					//		excess,
					//		BalanceStatus::Free,
					//	).defensive();
```